compose_box: Add start compose call buttons

MritunjayTiwari14 opened the pull request:
Fixes #1824 
Fixes #2062

Chat thread: 

- [#mobile-dev-help > unlike other video icon get filled with color property @ 💬](https://chat.zulip.org/#narrow/channel/516-mobile-dev-help/topic/unlike.20other.20video.20icon.20get.20filled.20with.20color.20property/near/2294564)
- [#frontend > zulip account session mismatch on flutter client during zoom](https://chat.zulip.org/#narrow/channel/6-frontend/topic/zulip.20account.20session.20mismatch.20on.20flutter.20client.20during.20zoom/with/2321288)
- [#mobile > general OAuth setup for zoom](https://chat.zulip.org/#narrow/channel/48-mobile/topic/general.20OAuth.20setup.20for.20zoom/with/2314781)
- [#api documentation > calls/zoom/create](https://chat.zulip.org/#narrow/channel/412-api-documentation/topic/calls.2Fzoom.2Fcreate/with/2328262)
- [#mobile-design > voice call button for zulip flutter app](https://chat.zulip.org/#narrow/channel/530-mobile-design/topic/voice.20call.20button.20for.20zulip.20flutter.20app/with/2349490)
- [#mobile-dev-help > zoom oauth issue](https://chat.zulip.org/#narrow/channel/516-mobile-dev-help/topic/zoom.20oauth.20issue/with/2323251)
- [#mobile-design > zoom oauth review](https://chat.zulip.org/#narrow/channel/530-mobile-design/topic/zoom.20oauth.20review/with/2416515)

### Design (Outdated)
| Before | After | After(Clicking Button) |
|---|---|---|
| ![b](https://github.com/user-attachments/assets/f8525bf8-5d2c-455e-9c43-d97b8c17edcc) | R![click_b](https://github.com/user-attachments/assets/f3fe4b37-c645-438b-9875-3308f18bdbe4) | ![b (2)](https://github.com/user-attachments/assets/57a5d34c-930b-4565-9943-d697160ac683) |
| ![w_b](https://github.com/user-attachments/assets/8c569516-0bf3-48d8-8276-cb0ebcede00b) | ![w_click_b](https://github.com/user-attachments/assets/0f636763-14db-44ee-a3cd-5d1fabc5909f) | ![w_n](https://github.com/user-attachments/assets/3c375b97-9c5d-4604-aa9b-5691c7a5166a) |




### Light Theme
https://github.com/user-attachments/assets/6b5af98e-4b28-44f4-b203-151d18f82897

### Dark Theme
https://github.com/user-attachments/assets/fa3692dc-27e5-4498-b5fc-548b247c17c6

MritunjayTiwari14 commented:
All the TODO are done @gnprice!

gnprice commented:
Thanks. Before this can be reviewed, you'll need to revise it to have clear and coherent commits. See the instructions we discussed on a previous PR thread: https://github.com/zulip/zulip-flutter/pull/1830#pullrequestreview-3297396734

MritunjayTiwari14 committed:
icons: Add 'video', from the Figma

Figma:
  https://www.figma.com/design/1JTNtYo9memgW7vV6d0ygq/Zulip-Mobile?node-id=3704-41538&t=udaDEYxIvoxfLwqQ-4

# Conflicts:
#	assets/icons/ZulipIcons.ttf

gnprice commented:
Thanks. Some of these commits aren't coherent. For example this commit f314a7444886b3c4ba3f4519ef9968efe3e94eb7:

```
    deps: Update pigeon to `26.0.2` from `26.0.1`
    
    This commit is the result of the following commands:
        flutter pub upgrade --major-version pigeon
        tools/check --all-files --fix pigeon
    
    Changelog:
    
    https://pub.dev/packages/pigeon/changelog#2602
---
 .../kotlin/com/zulip/flutter/AndroidIntents.g.kt    | Bin 6118 -> 6118 bytes
 .../com/zulip/flutter/AndroidNotifications.g.kt     | Bin 30643 -> 30643 bytes
 ios/Runner/Notifications.g.swift                    |   2 +-
 3 files changed, 1 insertion(+), 1 deletion(-)
```

can't be right, because it says it involved `flutter pub upgrade` but there's no change to pubspec.lock.

One useful step for spotting this kind of issue is to run our tests at every commit. If they fail, the commit needs fixing.

It's straightforward to do that by hand. But a command which can be a useful shortcut for it is this:
`git rebase --exec 'tools/check --diff @~'`

gnprice commented:
Chat thread: [#mobile-dev-help > unlike other video icon get filled with color property @ 💬](https://chat.zulip.org/#narrow/channel/516-mobile-dev-help/topic/unlike.20other.20video.20icon.20get.20filled.20with.20color.20property/near/2294564)

chrisbobbe commented:
> the issue has been resolved

Could you update the screenshots to show this please?

chrisbobbe commented on assets/l10n/app_en.arb:
```diff
@@ -574,6 +574,10 @@
+  "composeBoxAttachFromVideoCallTooltip": "Attach a video call",
+    "@composeBoxAttachFromVideoCallTooltip": {
+      "description": "Tooltip for compose box icon to attach a video call url to the message."
+    },
```
Also please fix the indentation so it's consistent with other entries here.

chrisbobbe reviewed (commented):
Thanks! Here's a review of the first three commits:

d90df436c icons: Add 'video', from the Figma
589895266 api: Add realmVideoChatProvider and jitsiServerUrl to InitialSnapshot
d5bbd96d3 realm: Add getters for jitsiServerUrl and realmVideoChatProvider

And a partial review of the fourth:

2ea860264 compose: Add video chat button to compose box

MritunjayTiwari14 committed:
api: Add realmVideoChatProvider to InitialSnapshot

MritunjayTiwari14 committed:
api: Add realmAvailableVideoChatProviders to InitialSnapshot

MritunjayTiwari14 committed:
api: Add hasZoomToken to InitialSnapshot

MritunjayTiwari14 committed:
api: Add jitsi server url configurations to InitialSnapshot

MritunjayTiwari14 committed:
realm: Add RealmStore.realmAvailableVideoChatProviders

MritunjayTiwari14 committed:
store: Add getter for hasZoomToken

chrisbobbe commented:
CI is failing; please investigate, fix, and comment when this is ready for another review.

chrisbobbe commented on lib/api/model/model.dart:
```diff
@@ -213,6 +213,28 @@ class RealmEmojiItem {
+  jitsi(apiValue: 1),
+  zoomUser(apiValue: 3),
+  bigBlueButton(apiValue: 4),
+  zoomServer(apiValue: 5),
```
In some of these names, let's match the API documentation a bit more: `jitsiMeet`, `zoomUserOAuth`, `zoomServerToServerOAuth`.

chrisbobbe commented on lib/api/model/initial_snapshot.dart:
```diff
@@ -98,6 +98,8 @@ class InitialSnapshot {
 
   final int realmVideoChatProvider;
 
+  final String? realmJitsiServerUrl;
```
nit:

```suggestion
  final String? realmJitsiServerUrl; // TODO(server-8)
```

chrisbobbe commented on lib/api/model/initial_snapshot.dart:
```diff
@@ -121,10 +123,14 @@ class InitialSnapshot {
 
   final Map<String, RealmDefaultExternalAccount> realmDefaultExternalAccounts;
 
+  final String? jitsiServerUrl;
```
```suggestion
  final String? jitsiServerUrl; // TODO(server-8): Can ignore this deprecated field
```

chrisbobbe commented on lib/api/model/initial_snapshot.dart:
```diff
@@ -121,10 +123,14 @@ class InitialSnapshot {
 
   final Uri serverEmojiDataUrl;
 
+  final String? serverJitsiServerUrl;
```
```suggestion
  final String? serverJitsiServerUrl; // TODO(server-8)
```

chrisbobbe commented on lib/model/realm.dart:
```diff
@@ -47,6 +47,10 @@ mixin RealmStore on PerAccountStoreBase, UserGroupStore {
   GroupSettingValue? get realmCanDeleteOwnMessageGroup; // TODO(server-10)
   bool get realmEnableReadReceipts;
   bool get realmMandatoryTopics;
+  int get realmVideoChatProvider;
```
This looks like the right section for `realmVideoChatProvider`; see the heading above:

```dart
  //|//////////////////////////////
  // Realm settings found in realm/update_dict events:
  //   https://zulip.com/api/get-events#realm-update_dict
  //
  // In order of appearance in the realm/update_dict event doc.
  // TODO(#668): update all these realm settings on events.
```

If you search for `video_chat_provider:` on https://zulip.com/api/get-events#realm-update_dict you'll find that it's documented on the `realm / update_dict` event, so it's right to include it here. It's in the wrong order, though; see "In order of…" in that header comment.

chrisbobbe commented on lib/api/model/model.dart:
```diff
@@ -76,6 +76,24 @@ class RealmAvailableVideoChatProviders {
+  });
+
+  factory VideoCallResponse.fromJson(Map<String, dynamic> json)  =>
+      _$VideoCallResponseFromJson(json);
```
nit:

```suggestion
  factory VideoCallResponse.fromJson(Map<String, dynamic> json) =>
    _$VideoCallResponseFromJson(json);
```

chrisbobbe commented on lib/api/route/video_call.dart:
```diff
@@ -0,0 +1,23 @@
+/// Creates a Zoom video/audio call
+///
+/// POST /api/v1/calls/zoom/create
+Future<VideoCallResponse> createZoomCall(ApiConnection connection, {
```
We should link to the API doc. I'm having trouble finding it, which is unusual—just opened a discussion on CZO to find out: [#api documentation > calls/zoom/create @ 💬](https://chat.zulip.org/#narrow/channel/412-api-documentation/topic/calls.2Fzoom.2Fcreate/near/2327854)

(Reminder to myself to re-review this after I've found the API doc, because I'll need to consult it 🙂.)

chrisbobbe reviewed (commented):
Thanks! Here's a review of the API/model changes:

1c07e3f4c icons: Add 'video', from the Figma
585d056c9 api: Add realmVideoChatProvider to InitialSnapshot
ac3c1c015 api: Add realmAvailableVideoChatProviders to InitialSnapshot
2bd72328e api: Add hasZoomToken to InitialSnapshot
3e8e1de25 api: Add jitsi server url configurations to InitialSnapshot
340df7080 realm: Add getters for realmAvailableVideoChatProviders
9c4632476 realm: Add getters for jitsiServerUrl and realmVideoChatProvider
37a9d94f6 store: Add getter for hasZoomToken
0671977c0 api: Add routes to fetch zoom and bigBlueButton call url
f35502ad4 api: Add event `hasZoomToken`

leaving the last commit for a later round of review:

3827b6258 compose_box: Add `Add video call` button

MritunjayTiwari14 committed:
realm: Add RealmStore.jitsiServerUrl and RealmStore.realmVideoChatProvider

Rearranged `realmEnableReadReceipts` in proper order under
class `RealmStoreImpl` which was out of order previously.

# Conflicts:
#	lib/model/realm.dart

MritunjayTiwari14 committed:
api: Add event `HasZoomTokenEvent`

MritunjayTiwari14 committed:
api: Add routes to fetch zoom and bigBlueButton call url

MritunjayTiwari14 commented:
I've also added test for Zoom video call insertion in the final commit. PTAL! @chrisbobbe

MritunjayTiwari14 committed:
compose_box: Add Jitsi Meet support in `_AddComposeCallUrlButton`

MritunjayTiwari14 committed:
compose_box: Add Big Blue Button support in `_AddComposeCallUrlButton`

MritunjayTiwari14 committed:
compose_box: Add Zoom support in `_AddComposeCallUrlButton`

MritunjayTiwari14 commented:
Thank you @chrisbobbe for the patience and advice. I have expanded the final commit into several parts for each meeting application configuration. PTAL!

MritunjayTiwari14 committed:
icons: Add 'voice', from the Figma

Figma:
  https://www.figma.com/design/jbNOiBWvbtLuHaiTj4CW0G/Zulip-Web-App?node-id=1095-67204&t=dNTmxz03TSxfEiwf-4

MritunjayTiwari14 committed:
compose_box: Configure voice call button

Fixes #2062.

MritunjayTiwari14 commented:
Thank you for the patience @chrisbobbe, i have cleared the conflicts with the base branch. Please take a look!

gnprice reviewed (commented):
Thanks @MritunjayTiwari14 for your work on this!

This isn't a full review — I'll leave that still for @chrisbobbe — but I took a quick look through the PR, and here are the things I initially noticed. I think several of these would help directly in making it easier to review the PR.

Two other comments, on the commit series:
* Take a look through the commit messages, with `git log`. There are a couple of them that have blocks like this:

      # Conflicts:
      #       assets/icons/ZulipIcons.ttf

which were reminders Git offered to you when resolving a merge/rebase conflict, and aren't relevant to a reader of the resulting commit.

* Take a look through an overview of the whole sequence of commits, with `git log --oneline` (e.g., `git log --oneline origin..`; or a command I have aliased and use regularly, `git log --no-decorate --oneline @{u}.. --reverse`). Here's what that latter command produces:

14076ad70 icons: Add 'video', from the Figma
26eb9c782 api: Add realmVideoChatProvider to InitialSnapshot
da96c8461 api: Add realmAvailableVideoChatProviders to InitialSnapshot
5104f8c3e api: Add hasZoomToken to InitialSnapshot
dd8f888e3 api: Add jitsi server url configurations to InitialSnapshot
058b45f77 realm: Add RealmStore.realmAvailableVideoChatProviders
1e1100e3d realm: Add RealmStore.jitsiServerUrl and RealmStore.realmVideoChatProvider
475ca6b24 store: Add getter for hasZoomToken
d68a0ec88 api: Add routes to fetch zoom and bigBlueButton call url
b480c7d44 api: Add event `HasZoomTokenEvent`
d673329ea compose_box: Add Jitsi Meet support in `_AddComposeCallUrlButton`
112d54615 compose_box: Add Big Blue Button support in `_AddComposeCallUrlButton`
ca366e8ad compose_box: Add Zoom support in `_AddComposeCallUrlButton`
dbedef951 icons: Add 'voice', from the Figma
dd937bd8e compose_box: Configure voice call button

The PR can be made easier to read and review by reordering those to have a bit more of a clear narrative. For example: first all the pieces needed for one provider, like Jitsi; then the additional pieces for another provider; then another.
